## **Capstone Project (IBM Applied Data Science) - Section 1**

### **Goal: Segmenting and Clustering Neighbourhoods of Toronto**

### In this notebook we are going to explore Toronto city by segmenting and clustering

### its neighbourhoods using Foursquare API.

## **Part 1**

In [74]:
# Libraries and settings

import numpy as np # To handle data in a vectorized manner
import pandas as pd # Data analsysis

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # To Handle JSON files

!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # To convert an address into latitude and longitude values

import requests # To Handle requests
from pandas.io.json import json_normalize # To tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium # Map rendering

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


### Let's get the Toronto Neighbourhoods' data by scraping a relevant wikipedia page<br> and transform it into a pandas dataframe:

In [4]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
dfT = pd.read_html(url)[0]
dfT

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
26          M9C         Etobicoke   
27          M1E       Scarborough   
28          M2E      Not assigned   
29          M3E      Not assigned   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
33          M7E      Not assigned   
34          M8E      Not assigned   
35          M9E      Not assigned   
36          M1G       Scarborough   
37          M2G      Not assigned   
38          M3G      Not assigned   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
42          M7G      Not assigned   
43          M8G      Not assigned   
44          M9G      Not assigned   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
51          M7H      Not assigned   
52          M8H      Not assigned   
53          M9H      Not assigned   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
60          M7J      Not assigned   
61          M8J      Not assigned   
62          M9J      Not assigned   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
69          M7K      Not assigned   
70          M8K      Not assigned   
71          M9K      Not assigned   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
78          M7L      Not assigned   
79          M8L      Not assigned   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
87          M7M      Not assigned   
88          M8M      Not assigned   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
96          M7N      Not assigned   
97          M8N      Not assigned   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
101         M3P      Not assigned   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
105         M7P      Not assigned   
106         M8P      Not assigned   
107     

### Checking how many of boroughs are "Not assigned":

In [6]:
dfT[dfT.Borough == "Not assigned"]

Postal Code       Borough Neighbourhood
0           M1A  Not assigned  Not assigned
1           M2A  Not assigned  Not assigned
7           M8A  Not assigned  Not assigned
10          M2B  Not assigned  Not assigned
15          M7B  Not assigned  Not assigned
16          M8B  Not assigned  Not assigned
19          M2C  Not assigned  Not assigned
24          M7C  Not assigned  Not assigned
25          M8C  Not assigned  Not assigned
28          M2E  Not assigned  Not assigned
29          M3E  Not assigned  Not assigned
33          M7E  Not assigned  Not assigned
34          M8E  Not assigned  Not assigned
35          M9E  Not assigned  Not assigned
37          M2G  Not assigned  Not assigned
38          M3G  Not assigned  Not assigned
42          M7G  Not assigned  Not assigned
43          M8G  Not assigned  Not assigned
44          M9G  Not assigned  Not assigned
51          M7H  Not assigned  Not assigned
52          M8H  Not assigned  Not assigned
53          M9H  Not assigned  Not assigned
60          M7J  Not assigned  Not assigned
61          M8J  Not assigned  Not assigned
62          M9J  Not assigned  Not assigned
69          M7K  Not assigned  Not assigned
70          M8K  Not assigned  Not assigned
71          M9K  Not assigned  Not assigned
78          M7L  Not assigned  Not assigned
79          M8L  Not assigned  Not assigned
87          M7M  Not assigned  Not assigned
88          M8M  Not assigned  Not assigned
96          M7N  Not assigned  Not assigned
97          M8N  Not assigned  Not assigned
101         M3P  Not assigned  Not assigned
105         M7P  Not assigned  Not assigned
106         M8P  Not assigned  Not assigned
110         M3R  Not assigned  Not assigned
115         M8R  Not assigned  Not assigned
118         M2S  Not assigned  Not assigned
119         M3S  Not assigned  Not assigned
123         M7S  Not assigned  Not assigned
124         M8S  Not assigned  Not assigned
125         M9S  Not assigned  Not assigned
127         M2T  Not assigned  Not assigned
128         M3T  Not assigned  Not assigned
131         M6T  Not assigned  Not assigned
132         M7T  Not assigned  Not assigned
133         M8T  Not assigned  Not assigned
134         M9T  Not assigned  Not assigned
136         M2V  Not assigned  Not assigned
137         M3V  Not assigned  Not assigned
140         M6V  Not assigned  Not assigned
141         M7V  Not assigned  Not assigned
145         M2W  Not assigned  Not assigned
146         M3W  Not assigned  Not assigned
149         M6W  Not assigned  Not assigned
150         M7W  Not assigned  Not assigned
154         M2X  Not assigned  Not assigned
155         M3X  Not assigned  Not assigned
158         M6X  Not assigned  Not assigned
159         M7X  Not assigned  Not assigned
161         M9X  Not assigned  Not assigned
162         M1Y  Not assigned  Not assigned
163         M2Y  Not assigned  Not assigned
164         M3Y  Not assigned  Not assigned
166         M5Y  Not assigned  Not assigned
167         M6Y  Not assigned  Not assigned
170         M9Y  Not assigned  Not assigned
171         M1Z  Not assigned  Not assigned
172         M2Z  Not assigned  Not assigned
173         M3Z  Not assigned  Not assigned
174         M4Z  Not assigned  Not assigned
175         M5Z  Not assigned  Not assigned
176         M6Z  Not assigned  Not assigned
177         M7Z  Not assigned  Not assigned
179         M9Z  Not assigned  Not assigned

### Wrangling and cleansing the dataframe:

In [18]:
dfTo = dfT[dfT.Borough != "Not assigned"]
dfTo

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
8           M9A         Etobicoke   
9           M1B       Scarborough   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
17          M9B         Etobicoke   
18          M1C       Scarborough   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
26          M9C         Etobicoke   
27          M1E       Scarborough   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
36          M1G       Scarborough   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
107         M9P         Etobicoke   
108         M1R       Scarborough   
109         M2R        North York   
111         M4R   Central Toronto   
112         M5R   Central Toronto   
113         M6R      West Toronto   
114         M7R       Mississauga   
116         M9R         Etobicoke   
117         M1S       Scarborough   
120         M4S   Central Toronto   
121         M5S  Downtown Toronto   
122         M6S      West Toronto   
126         M1T       Scarborough   
129         M4T   Central Toronto   
130         M5T  Downtown Toronto   
135         M1V       Scarborough   
138         M4V   Central Toronto   
139         M5V  Downtown Toronto   
142         M8V         Etobicoke   
143         M9V         Etobicoke   
144         M1W       Scarborough   
147         M4W  Downtown Toronto   
148         M5W  Downtown Toronto   
151         M8W         Etobicoke   
152         M9W         Etobicoke   
153         M1X       Scarborough   
156         M4X  Downtown Toronto   
157         M5X  Downtown Toronto   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Vic

In [20]:
dfTo = dfTo.reset_index()
dfTo.drop(['index'], axis=1, inplace=True)

In [21]:
dfTo

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Vic

## We are going to work on 103 different postal codes including <br>their Boroughs and Neighbourhoods of Toronto city.

In [22]:
dfTo.shape

(103, 3)

In [23]:
!pip install geopandas
!pip install geopy

## **Part 2**

### Getting the Latitude and Longitude of each postal code:

In [91]:
url = "http://cocl.us/Geospatial_data"
postcode = pd.read_csv(url)
postcode

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
30          M3K  43.737473 -79.464763
31          M3L  43.739015 -79.506944
32          M3M  43.728496 -79.495697
33          M3N  43.761631 -79.520999
34          M4A  43.725882 -79.315572
35          M4B  43.706397 -79.309937
36          M4C  43.695344 -79.318389
37          M4E  43.676357 -79.293031
38          M4G  43.709060 -79.363452
39          M4H  43.705369 -79.349372
40          M4J  43.685347 -79.338106
41          M4K  43.679557 -79.352188
42          M4L  43.668999 -79.315572
43          M4M  43.659526 -79.340923
44          M4N  43.728020 -79.388790
45          M4P  43.712751 -79.390197
46          M4R  43.715383 -79.405678
47          M4S  43.704324 -79.388790
48          M4T  43.689574 -79.383160
49          M4V  43.686412 -79.400049
50          M4W  43.679563 -79.377529
51          M4X  43.667967 -79.367675
52          M4Y  43.665860 -79.383160
53          M5A  43.654260 -79.360636
54          M5B  43.657162 -79.378937
55          M5C  43.651494 -79.375418
56          M5E  43.644771 -79.373306
57          M5G  43.657952 -79.387383
58          M5H  43.650571 -79.384568
59          M5J  43.640816 -79.381752
60          M5K  43.647177 -79.381576
61          M5L  43.648198 -79.379817
62          M5M  43.733283 -79.419750
63          M5N  43.711695 -79.416936
64          M5P  43.696948 -79.411307
65          M5R  43.672710 -79.405678
66          M5S  43.662696 -79.400049
67          M5T  43.653206 -79.400049
68          M5V  43.628947 -79.394420
69          M5W  43.646435 -79.374846
70          M5X  43.648429 -79.382280
71          M6A  43.718518 -79.464763
72          M6B  43.709577 -79.445073
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

### Merging neighbourhoods of Toronto with Latitude and Longitude dataframe:

In [25]:
df_toronto = dfTo.merge(postcode, how = 'inner', on = ['Postal Code'])
df_toronto

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656

## **Part 3**

## Explore and cluster the neighbourhoods of Toronto.

### Getting the latitude and longitude values(coordinates) of Toronto by using geopy library:

In [26]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="ny_explorer")

location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude

print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


### Visualizing Toronto's map with neighbourhoods superimposed on it.

### We could zoom into the below map:

In [27]:
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for lat, lng, label in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

### Defining credintials to use Foursquare API:

In [76]:
CLIENT_ID = 'SGKLM1LRLIJIJ1YLEHCL5A4HHECUTS1SHIIEQKJQVYEYRNAK' # Foursquare ID
CLIENT_SECRET = 'GSULTSI4CI2NEF5H2VURC4IZZBQXFQEWUXJLE044HLVT2DZI' # Foursquare Secret
VERSION = '20200806' # Foursquare API version

print('Credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Credentails:
CLIENT_ID: SGKLM1LRLIJIJ1YLEHCL5A4HHECUTS1SHIIEQKJQVYEYRNAK
CLIENT_SECRET:GSULTSI4CI2NEF5H2VURC4IZZBQXFQEWUXJLE044HLVT2DZI


In [29]:
Neighbourhood_latitude = df_toronto.loc[0, 'Latitude'] # neighborhood latitude value
Neighbourhood_longitude = df_toronto.loc[0, 'Longitude'] # neighborhood longitude value

Neighbourhood_name = df_toronto.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(Neighbourhood_name, 
                                                               Neighbourhood_latitude, 
                                                               Neighbourhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


### Now, we are ready to utilize Foursquare API in order to explore and segment Toronto's neighbourhoods:

In [30]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
 
radius = 500 # defining radius
 
# creating URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    Neighbourhood_latitude, 
    Neighbourhood_longitude, 
    radius, 
    LIMIT)
url # displaying URL

'https://api.foursquare.com/v2/venues/explore?&client_id=SGKLM1LRLIJIJ1YLEHCL5A4HHECUTS1SHIIEQKJQVYEYRNAK&client_secret=GSULTSI4CI2NEF5H2VURC4IZZBQXFQEWUXJLE044HLVT2DZI&v=20200806&ll=43.7532586,-79.3296565&radius=500&limit=100'

In [31]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5f2c58f7be3b4d177927e836'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 2,
  'suggestedBounds': {'ne': {'lat': 43.757758604500005,
    'lng': -79.32343823984928},
   'sw': {'lat': 43.7487585955, 'lng': -79.33587476015072}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': 

In [32]:
# function for extracting the venue's category
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

### Cleaning the json result and transform it into a pandas dataframe:

In [77]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flattening JSON

# filtering columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filtering the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# cleaning columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/Users/jayexoplanet/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name         categories        lat        lng
0  Brookbanks Park               Park  43.751976 -79.332140
1    Variety Store  Food & Drink Shop  43.751974 -79.333114

## **Exploring Neighbourhoods in Toronto city**
### Creating a function to do the same process on venues for all neighbourhoods:

In [34]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # creating the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # making the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # returning only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [35]:
Toronto_venues = getNearbyVenues(names=df_toronto['Neighbourhood'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [36]:
print(Toronto_venues.shape)
Toronto_venues.head()

(2134, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3               Portugril       43.725819       -79.312785   
4             Tim Hortons       43.725517       -79.313103   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3  Portuguese Restaurant  
4            Coffee Shop

### Number of venues for each neighbourhood:

In [37]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               5   
Alderwood, Long Branch                                                  7   
Bathurst Manor, Wilson Heights, Downsview North                        21   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      24   
Berczy Park                                                            57   
Birch Cliff, Cliffside West                                             5   
Brockton, Parkdale Village, Exhibition Place                           24   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  12   
Cedarbrae                                                               8   
Central Bay Street                                                     66   
Christie                                                               17   
Church and Wellesley                                                   75   
Clarks Corners, Tam O'Shanter, Sullivan                                14   
Cliffside, Cliffcrest, Scarborough Village West                         2   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             32   
Davisville North                                                       10   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        5   
Don Mills                                                              25   
Dorset Park, Wexford Heights, Scarborough Town ...                      6   
Downsview                                                              13   
Dufferin, Dovercourt Village                                           14   
East Toronto, Broadview North (Old East York)                           3   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      9   
Fairview, Henry Farm, Oriole                                           66   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Glencairn                                                               4   
Golden Mile, Clairlea, Oakridge                                         9   
Guildwood, Morningside, West Hill                                       8   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
Hillcrest Village                                                       4   
Humber Summit                                                           2   
Humberlea, Emery                                                        1   
Humewood-Cedarvale                                                      3   
India Bazaar, The Beaches West                                         21   
Islington Avenue, Humber Valley Village                                 1   
Kennedy Park, Ionview, East Birchmount Park                             4   
Kensington Market, Chinatown, Grange Park                              65   
Kingsview Village, St. Phillips, Martin Grove G...                      4   
Lawrence Manor, Lawrence Heights                                       16   
Lawrence Park                                                           3   
Leaside                                                                33   
Little Portugal, Trinity                                               47   
Malvern, Rouge                                  

In [38]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 268 uniques categories.


### Analyzing neighbourhoods by One Hot encoding:

In [42]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# adding neighbourhood column back to dataframe
Toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

# moving neighbourhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Garage  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0            0              0          0           0           0       0   
1            0              0          0           0           0       0   
2            0              0          0           0           0       0   
3            0              0          0           0           0       0   
4            0              0          0           0           0       0   

   Bank  Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0     0    0               0                 0                 0   
1     0    0               0                 0                 0   
2     0    0               0                 0                 0   
3     0    0               0                 0                 0   
4     0    0               0                 0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0          0       0              0          0         0   
1                   0          0       0              0          0         0   
2                   0          0       0              0          0         0   
3                   0          0       0              0          0         0   
4                   0          0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \

### Grouping rows by neighbourhood and taking the mean of the frequency of each category:

In [45]:
Toronto_onehot = Toronto_onehot.groupby('Neighborhood').mean().round(2).reset_index()
Toronto_onehot

Neighborhood  Yoga Studio  \
0                                           Agincourt         0.00   
1                              Alderwood, Long Branch         0.00   
2     Bathurst Manor, Wilson Heights, Downsview North         0.00   
3                                     Bayview Village         0.00   
4                   Bedford Park, Lawrence Manor East         0.00   
5                                         Berczy Park         0.00   
6                         Birch Cliff, Cliffside West         0.00   
7        Brockton, Parkdale Village, Exhibition Place         0.04   
8   Business reply mail Processing Centre, South C...         0.06   
9   CN Tower, King and Spadina, Railway Lands, Har...         0.00   
10                                Caledonia-Fairbanks         0.00   
11              Canada Post Gateway Processing Centre         0.00   
12                                          Cedarbrae         0.00   
13                                 Central Bay Street         0.02   
14                                           Christie         0.00   
15                               Church and Wellesley         0.03   
16            Clarks Corners, Tam O'Shanter, Sullivan         0.00   
17    Cliffside, Cliffcrest, Scarborough Village West         0.00   
18                     Commerce Court, Victoria Hotel         0.00   
19                                         Davisville         0.00   
20                                   Davisville North         0.00   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn         0.00   
22                                          Don Mills         0.00   
23  Dorset Park, Wexford Heights, Scarborough Town...         0.00   
24                                          Downsview         0.00   
25                       Dufferin, Dovercourt Village         0.00   
26      East Toronto, Broadview North (Old East York)         0.00   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...         0.00   
28                       Fairview, Henry Farm, Oriole         0.00   
29             First Canadian Place, Underground city         0.00   
30    Forest Hill North & West, Forest Hill Road Park         0.00   
31                           Garden District, Ryerson         0.00   
32                                          Glencairn         0.00   
33                    Golden Mile, Clairlea, Oakridge         0.00   
34                  Guildwood, Morningside, West Hill         0.00   
35  Harbourfront East, Union Station, Toronto Islands         0.00   
36                      High Park, The Junction South         0.00   
37                                  Hillcrest Village         0.00   
38                                      Humber Summit         0.00   
39                                   Humberlea, Emery         0.00   
40                                 Humewood-Cedarvale         0.00   
41                     India Bazaar, The Beaches West         0.00   
42            Islington Avenue, Humber Valley Village         0.00   
43        Kennedy Park, Ionview, East Birchmount Park         0.00   
44          Kensington Market, Chinatown, Grange Park         0.00   
45  Kingsview Village, St. Phillips, Martin Grove ...         0.00   
46                   Lawrence Manor, Lawrence Heights         0.00   
47                                      Lawrence Park         0.00   
48                                            Leaside         0.00   
49                           Little Portugal, Trinity         0.02   
50                                     Malvern, Rouge         0.00   
51  Milliken, Agincourt North, Steeles East, L'Amo...         0.00   
52  Mimico NW, The Queensway West, South of Bloor,...         0.00   
53                        Moore Park, Summerhill East         0.00   
54       New Toronto, Mimico South, Humber Bay Shores         0.00   
55           North Park, Maple Leaf Park, Upwood Park         0.00   
56                  North Toronto West, Lawrence Pa

### Neighbourhoods of Toronto with the top 5 most common venues:

In [47]:
num_top_venues = 5

for hood in Toronto_onehot['Neighborhood']:
    print("----"+hood+"----")
    temp = Toronto_onehot[Toronto_onehot['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0               Skating Rink   0.2
1  Latin American Restaurant   0.2
2                     Lounge   0.2
3             Clothing Store   0.2
4             Breakfast Spot   0.2


----Alderwood, Long Branch----
            venue  freq
0     Pizza Place  0.29
1             Gym  0.14
2     Coffee Shop  0.14
3  Sandwich Place  0.14
4             Pub  0.14


----Bathurst Manor, Wilson Heights, Downsview North----
              venue  freq
0              Bank  0.10
1       Coffee Shop  0.10
2  Sushi Restaurant  0.05
3       Gas Station  0.05
4       Bridal Shop  0.05


----Bayview Village----
                 venue  freq
0                 Bank  0.25
1   Chinese Restaurant  0.25
2  Japanese Restaurant  0.25
3                 Café  0.25
4          Yoga Studio  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0         Coffee Shop  0.08
1          Restaurant  0.08
2      Sandwich Place  0.08
3  Italian Restaurant  0.

### Creating a function for sorting the venues:

In [48]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

### Top 10 venues for each neighborhood:

In [49]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_onehot['Neighborhood']

for ind in np.arange(Toronto_onehot.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_onehot.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt          Skating Rink   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village    Chinese Restaurant   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Clothing Store                Lounge   
1               Skating Rink                   Gym                   Pub   
2                       Bank        Sandwich Place     Mobile Phone Shop   
3                       Café   Japanese Restaurant                  Bank   
4                Coffee Shop            Restaurant        Sandwich Place   

  5th Most Common Venue    6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot      Distribution Center      Department Store   
1        Sandwich Place              Coffee Shop               Gay Bar   
2        Ice Cream Shop              Supermarket         Deli / Bodega   
3               Dog Run       Dim Sum Restaurant                 Diner   
4      Sushi Restaurant  Comfort Food Restaurant              Pharmacy   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0          Dessert Shop    Dim Sum Restaurant                  Diner  
1             Gastropub      Doner Restaurant                Dog Run  
2      Sushi Restaurant           Pizza Place               Pharmacy  
3        Discount Store   Distribution Center          Women's Store  
4           Pizza Place                  Café                Butcher

## **Clustering neighbourhoods**

In [71]:
# adding clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels2', kmeans.labels_)

toronto_merged = df_toronto

# merging toronto_grouped with toronto_data to add latitude/longitude for each neighbourhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.dropna(subset=['Cluster Labels2'], inplace=True)
toronto_merged['Cluster Labels2']=toronto_merged['Cluster Labels2'].astype('int32')
toronto_merged.head()


Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels2  Cluster Labels1  Cluster Labels  \
0  43.753259 -79.329656                1              1.0             1.0   
1  43.725882 -79.315572                1              1.0             1.0   
2  43.654260 -79.360636                1              1.0             1.0   
3  43.718518 -79.464763                1              1.0             1.0   
4  43.662301 -79.389494                1              1.0             1.0   

  1st Most Common Venue   2nd Most Common Venue       3rd Most Common Venue  \
0                  Park       Food & Drink Shop               Women's Store   
1           Coffee Shop             Pizza Place  Financial or Legal Service   
2           Coffee Shop                    Park                         Pub   
3        Clothing Store  Furniture / Home Store           Accessories Store   
4           Coffee Shop                   Diner                 Yoga Studio   

  4th Most Common Venue  5th Most Common Venue 6th Most Common Venue  \
0   Distribution Center       Department Store          Dessert Shop   
1          Hockey Arena  Portuguese Restaurant     French Restaurant   
2                Bakery                Theater        Breakfast Spot   
3         Women's Store               Boutique          Carpet Store   
4    Mexican Restaurant    Japanese Restaurant   Distribution Center   

  7th Most Common Venue 8th Most Common Venue  9th Most Common Venue  \
0    Dim Sum Restaurant                 Diner         Discount Store   
1            Comic Shop          Dance Studio              Drugstore   
2                  Café                   Spa         Ice Cream Shop   
3    Athletics & Sports           Event Space  Vietnamese Restaurant   
4         Burrito Place              Beer Bar          Smoothie Shop   

  10th Most Common Venue  
0                Dog Run  
1             Donut Shop  
2         Farmers Market  
3            Coffee Shop  
4  Portuguese Restaurant

In [57]:
df_toronto = df_toronto.rename(columns={'Neighbourhood': 'Neighborhood'})

df_toronto

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656

### **Plotting and visualizing the clusters as the desired result**

In [72]:
# creating map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# setting color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# adding markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels2']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## **Examining clusters**

### Cluster 1

In [82]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels2  Cluster Labels1  Cluster Labels  \
6  Scarborough                0              0.0             0.0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
6  Fast Food Restaurant         Women's Store          Dance Studio   

         4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
6  Eastern European Restaurant   Dumpling Restaurant             Drugstore   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
6            Donut Shop      Doner Restaurant               Dog Run   

  10th Most Common Venue  
6    Distribution Center

### Cluster 2
### We observe that most of venues are concentrated around Downtown Toronto.

In [84]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels2  Cluster Labels1  Cluster Labels  \
0          North York                1              1.0             1.0   
1          North York                1              1.0             1.0   
2    Downtown Toronto                1              1.0             1.0   
3          North York                1              1.0             1.0   
4    Downtown Toronto                1              1.0             1.0   
5           Etobicoke                1              1.0             1.0   
7          North York                1              1.0             1.0   
8           East York                1              1.0             1.0   
9    Downtown Toronto                1              1.0             1.0   
10         North York                1              1.0             1.0   
12        Scarborough                1              1.0             1.0   
13         North York                1              1.0             1.0   
14          East York                1              1.0             1.0   
15   Downtown Toronto                1              1.0             1.0   
16               York                1              1.0             1.0   
17          Etobicoke                1              1.0             1.0   
18        Scarborough                1              1.0             1.0   
19       East Toronto                1              1.0             1.0   
20   Downtown Toronto                1              1.0             1.0   
21               York                1              1.0             1.0   
22        Scarborough                1              1.0             1.0   
23          East York                1              1.0             1.0   
24   Downtown Toronto                1              1.0             1.0   
25   Downtown Toronto                1              1.0             1.0   
26        Scarborough                1              1.0             1.0   
27         North York                1              1.0             1.0   
28         North York                1              1.0             1.0   
29          East York                1              1.0             1.0   
30   Downtown Toronto                1              1.0             1.0   
31       West Toronto                1              1.0             1.0   
32        Scarborough                1              1.0             1.0   
33         North York                1              1.0             1.0   
34         North York                1              1.0             1.0   
35          East York                1              1.0             1.0   
36   Downtown Toronto                1              1.0             1.0   
37       West Toronto                1              1.0             1.0   
38        Scarborough                1              1.0             1.0   
39         North York                1              1.0             1.0   
40         North York                1              1.0             1.0   
41       East Toronto                1              1.0             1.0   
42   Downtown Toronto                1              1.0             1.0   
43       West Toronto                1              1.0             1.0   
44        Scarborough                1              1.0             1.0   
46         North York                1              1.0             1.0   
47       East Toronto                1              1.0             1.0   
48   Downtown Toronto                1              1.0             1.0   
49         North York                1              1.0             1.0   
50         North York                1              1.0             1.0   
51        Scarborough                1              1.0             1.0   
53         North York                1              1.0             1.0   
54       East Toronto                1              1.0             1.0   
55         North York                1              1.0             1.0   
56               York                1 

### Cluster 3

In [85]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels2  Cluster Labels1  Cluster Labels  \
57   North York                2              2.0             2.0   
101   Etobicoke                2              2.0             2.0   

    1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
57         Baseball Field               Dog Run      Department Store   
101        Baseball Field         Deli / Bodega     Electronics Store   

           4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
57                  Dessert Shop    Dim Sum Restaurant                 Diner   
101  Eastern European Restaurant   Dumpling Restaurant             Drugstore   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
57         Discount Store   Distribution Center         Women's Store   
101            Donut Shop      Doner Restaurant           Event Space   

    10th Most Common Venue  
57             Event Space  
101                Dog Run

### Cluster 4

In [87]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels2  Cluster Labels1  Cluster Labels  \
83  Central Toronto                3              3.0             3.0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
83            Restaurant         Women's Store          Dance Studio   

          4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
83  Eastern European Restaurant   Dumpling Restaurant             Drugstore   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
83            Donut Shop      Doner Restaurant               Dog Run   

   10th Most Common Venue  
83    Distribution Center

### Cluster 5

In [89]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels2  Cluster Labels1  Cluster Labels  \
11  Etobicoke                4              4.0             4.0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
11   Filipino Restaurant         Women's Store          Dance Studio   

          4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
11  Eastern European Restaurant   Dumpling Restaurant             Drugstore   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
11            Donut Shop      Doner Restaurant               Dog Run   

   10th Most Common Venue  
11    Distribution Center

#### **Created by J.E.**